Provide qemu-kvm process inbound/outbound connectivity.

vefimova opened the pull request:
Now process is spawned inside virtlet's netns and receives opened tap device FD to set up netdev.
Fixes #240

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mirantis/virtlet/248)
<!-- Reviewable:end -->

jellonek commented:
Initial comments leaved, will check this deeper again tomorrow.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[test.sh, line 17 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhODaH3PKqq2Nt3Q5Fx:-KhODaH3PKqq2Nt3Q5Fy:b-rdd0is) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/test.sh#L17)):*
> ```Shell
> #build/cmd.sh build
> #build/cmd.sh copy
> #build/cmd.sh test
> ```

Please remove these commented lines it they are not necessary there, or uncomment them if they should be there.

---

*[cmd/vmwrapper/vmwrapper.go, line 235 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhOEAmWFFJ0wUeid0XV:-KhOEAmWFFJ0wUeid0XW:b-29edj) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/cmd/vmwrapper/vmwrapper.go#L235)):*
> ```Go
> 	virtletNS, err := ns.GetCurrentNS()
> 	if err != nil {
> 		glog.Errorf("Failed to get current naet namespace: %v", err)
> ```

Typo. "naet"

---

*[cmd/vmwrapper/vmwrapper.go, line 280 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhOEgXBUK3SIPvf52mV:-KhOEgXBUK3SIPvf52mW:b-98afzc) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/cmd/vmwrapper/vmwrapper.go#L280)):*
> ```Go
> 			return err
> 		}); err != nil {
> 			glog.Error("Error occurred while running VM network namespace: %s", err)
> ```

... while starting subprocess in Virtlet base network namespace

---

*[docs/networking.md, line 59 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhOF-nHbtbnUwrhxi7u:-KhOF-nIeyWBiBCUkUmi:b-2qh8a2) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/docs/networking.md#L59)):*
> ```Markdown
> - VIRTLET_CNI_CONFIG - json sting with CNI settings, ex: {"ip4":{"ip":"10.1.91.2/24","gateway":"10.1.91.1","routes":[{"dst":"0.0.0.0/0"}]},"dns":{}}
> ```
>  - On CreateContainer request virlet calls libvirt api to start domain, which in its turn leads to running VMwrapper with all qemu args. Using info from set env vars VMWrapper sets up networking and runs dhcp-server inside pod's netns. VM itself is started inside virtlet's netns to provide inbound and outbound connectivity for qemu-kvm process.
> ```

qemu-kvm running VM itself is started inside virtlet's base netns to provide inbound and outbound connectivity.

---

*[docs/networking.md, line 65 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhOF_aEsau-82iBtufv:-KhOF_aEsau-82iBtufw:b30r82j) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/docs/networking.md#L65)):*
> ```Markdown
>     3. runs dhcp-server to pass ip to VM's eth stipped from veth and default routes
> 
> And at last VMWrapper spawns qemu-kvm inside virtlet's netns and provides it with FD of opened tap0.
> ```

Finally `vmwrapper` spawns qemu-kvm inside original virtlet's netns and passed to it file descriptor of opened tap0 device.

---

*[pkg/nettools/nettools.go, line 64 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhOGKf14DBOp7xSNVXd:-KhOGKf14DBOp7xSNVXe:b-ebc1ke) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/pkg/nettools/nettools.go#L64)):*
> ```Go
> )
> 
> // Had to duplicate ifReq here as it's not  exported
> ```

Double space.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/248#-:-KhOGmptoeg2oXjhRer7:bobexka)*
<!-- Sent from Reviewable.io -->

ivan4th commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[docs/networking.md, line 18 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhXGL1HFgmuyGFciAT2:-KhXGL1HFgmuyGFciAT3:b-5fv33m) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/docs/networking.md#L18)):*
> ```Markdown
> |               +---||--------------+                                                                                                                          |
> |                   \/                                                                                                                                         |
> |               FD of opened tap deivce                                                                                                                        |
> ```

FD of the tap device

---

*[docs/networking.md, line 61 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhXGF8nTLmmlsU6r1Vc:-KhXGF8nTLmmlsU6r1Vd:bmwytmg) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/docs/networking.md#L61)):*
> ```Markdown
>  - On CreateContainer request virlet calls libvirt api to start domain, which in its turn leads to running VMwrapper with all qemu args. Using info from set env vars VMWrapper sets up networking and runs dhcp-server inside pod's netns. VM itself is started inside virtlet's netns to provide inbound and outbound connectivity for qemu-kvm process.
> In more details, VMWrapper inside pod's netns performs the following:
>     1. creates tap0 for domain and br0, strips ip from veth end inside netns
> ```

strips IP address and routes

---

*[docs/networking.md, line 65 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhOF_aEsau-82iBtufv:-KhXFiSe9HjknecFitAu:bic2sxf) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/docs/networking.md#L65)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Finally `vmwrapper` spawns qemu-kvm inside original virtlet's netns and passed to it file descriptor of opened tap0 device.
</blockquote></details>

Finally vmwrapper spawns qemu-kvm (or qemu, if KVM is disabled) inside original virtlet's netns and passes it the file descriptor corresponding to tap0 device.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/248)*
<!-- Sent from Reviewable.io -->

ivan4th commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[test.sh, line 17 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhODaH3PKqq2Nt3Q5Fx:-KhXGiIh-l12GXTy22Ro:b-rs2kh8) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/test.sh#L17)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Please remove these commented lines it they are not necessary there, or uncomment them if they should be there.
</blockquote></details>

This breaks the build, please uncomment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/248)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Fixed comments, CI passes. 
Also rolled back workaroud for rbd dicsk attaching and added taking screenshots of vnc console to check its availablity.

---

Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[cmd/vmwrapper/vmwrapper.go, line 235 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhOEAmWFFJ0wUeid0XV:-KhcbBC3_4Uy4kbzCHFK:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/cmd/vmwrapper/vmwrapper.go#L235)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Typo. "naet"
</blockquote></details>

Done.

---

*[cmd/vmwrapper/vmwrapper.go, line 280 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhOEgXBUK3SIPvf52mV:-KhcbAjTF8BdLNGKsGZB:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/cmd/vmwrapper/vmwrapper.go#L280)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

... while starting subprocess in Virtlet base network namespace
</blockquote></details>

Done.

---

*[docs/networking.md, line 18 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhXGL1HFgmuyGFciAT2:-Khcb9mYT8yVy75OXxM_:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/docs/networking.md#L18)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

FD of the tap device
</blockquote></details>

Done.

---

*[docs/networking.md, line 59 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhOF-nHbtbnUwrhxi7u:-KhcbA3aX59N_d5RiCdt:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/docs/networking.md#L59)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

qemu-kvm running VM itself is started inside virtlet's base netns to provide inbound and outbound connectivity.
</blockquote></details>

Done.

---

*[docs/networking.md, line 61 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhXGF8nTLmmlsU6r1Vc:-KhcbCnGrIQdwFDFoqsG:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/docs/networking.md#L61)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

strips IP address and routes
</blockquote></details>

Done.

---

*[docs/networking.md, line 65 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhOF_aEsau-82iBtufv:-KhcbDDlI5a6-MFbWgRv:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/docs/networking.md#L65)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

Finally vmwrapper spawns qemu-kvm (or qemu, if KVM is disabled) inside original virtlet's netns and passes it the file descriptor corresponding to tap0 device.
</blockquote></details>

Done.

---

*[pkg/nettools/nettools.go, line 64 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhOGKf14DBOp7xSNVXd:-KhcbDoUhzhjkUL6H49l:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/pkg/nettools/nettools.go#L64)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Double space.
</blockquote></details>

Done.

---

*[test.sh, line 17 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/248#-KhODaH3PKqq2Nt3Q5Fx:-KhcbFuWE12FkrdFMZln:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/8ba34275bbdb43fba8761caa07c2a42fbc6b0d79/test.sh#L17)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

This breaks the build, please uncomment
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/248#-:-KhcbHuaLOBw8XpOnZtN:b894dc6)*
<!-- Sent from Reviewable.io -->

ivan4th commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/248#-:-KhccGSkA_wU-2YZGHPK:bnfp4nl)*
<!-- Sent from Reviewable.io -->
